### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by calmh**
I did the review here as a PR, to get somewhere we can comment and stuff. Will be adding commits for the subpackages.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -39,6 +39,8 @@ func main() {
 	flag.DurationVar(&pingInterval, "ping-interval", time.Minute, "How often pings are sent")
 	flag.DurationVar(&messageTimeout, "message-timeout", time.Minute, "Maximum amount of time we wait for relevant messages to arrive")
 
+	flag.Parse()
```
Comment (calmh) [inline main.go]: Needs to happen before we look at `extAddress`

File: protocol_listener.go [hunk 2]
```diff
@@ -35,6 +34,8 @@ func protocolListener(addr string, config *tls.Config) {
 			continue
 		}
 
+		setTCPOptions(conn)
```
Comment (calmh) [inline protocol_listener.go:37]: Moving to avoid potentially nil `conn` (on error).

File: session_listener.go [hunk 3]
```diff
@@ -35,10 +36,17 @@ func sessionListener(addr string) {
 }
 
 func sessionConnectionHandler(conn net.Conn) {
-	conn.SetDeadline(time.Now().Add(messageTimeout))
+	defer conn.Close()
+
+	if err := conn.SetDeadline(time.Now().Add(messageTimeout)); err != nil {
```
Comment (calmh) [inline session_listener.go:41]: Added error checking on the deadline stuff, although I don't really expect there to be an error here...

**Comment by calmh** [comment]
No comments on the `protocol` package

**Comment by calmh** [comment]
Did an `s/panic/log-and-return/` on client

**Comment by AudriusButkevicius** [comment]
Thanks, looks good.

**Pull request merged**